Test `test_broken_projections/test.py::test_broken_projections_in_backups_1` is flaky

[cidb](https://play.clickhouse.com/play?user=play#CldJVEgKICAgIGZhbHNlIGFzIG9ubHlfbWFzdGVyClNFTEVDVCAqCkZST00gKApTRUxFQ1QKICAgIHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHJudW0sCiAgICBzdWJzdHIoY29tbWl0X3NoYSwgMSw5KSBhcyBzaGEsCiAgICByb3VuZCh0ZXN0X2R1cmF0aW9uX21zIC8gMTAwMCwgMikgYXMgZHVyLAogICAgY2hlY2tfc3RhcnRfdGltZSwKICAgIGZvcm1hdFJlYWRhYmxlVGltZURlbHRhKGludERpdihub3coKSAtIGNoZWNrX3N0YXJ0X3RpbWUsIDM2MDApICogMzYwMCkgYXMgYWdvLAogICAgY2hlY2tfbmFtZSwKICAgIHRlc3RfbmFtZSwKICAgIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgMQogICAgQU5EIGNoZWNrX3N0YXJ0X3RpbWUgPj0gbm93KCkgLSBJTlRFUlZBTCAyOCBEQVkKICAgIEFORCAocHVsbF9yZXF1ZXN0X251bWJlciA9IDAgT1Igbm90IG9ubHlfbWFzdGVyKQogICAgQU5EIHRlc3Rfc3RhdHVzICE9ICdTS0lQUEVEJwogICAgQU5EIHRlc3Rfc3RhdHVzIExJS0UgJ0YlJwogICAgQU5EIGNoZWNrX3N0YXR1cyAhPSAnc3VjY2VzcycKICAgIEFORCBjaGVja19uYW1lIGxpa2UgJyVhYXIlJwopCldIRVJFIHRlc3RfbmFtZSBsaWtlICcldGVzdF9icm9rZW5fcHJvamVjdGlvbnNfaW5fYmFja3VwcyUnCk9SREVSIEJZIGNoZWNrX3N0YXJ0X3RpbWUgREVTQywgY2hlY2tfbmFtZSwgdGVzdF9uYW1l)

<details>
<summary>cidb</summary>



| prnum | sha | dur | check_start_time | ago | check_name | test_name | report_url |
|-:|:-|-:|-:|:-|:-|:-|:-|
| 55051 | b732723c1 | 2.73 | 2024-07-19 12:35:51 | 2 days and 21 hours | Integration tests (aarch64) [1/6] | test_broken_projections/test.py::test_broken_projections_in_backups_1 | https://s3.amazonaws.com/clickhouse-test-reports/55051/b732723c18204b4c1047c4021c9ea8ea20ef01ee/integration_tests__aarch64__[1_6].html |
| 66344 | 1c92e0933 | 2.48 | 2024-07-14 06:49:49 | 8 days and 3 hours | Integration tests (aarch64) [1/6] | test_broken_projections/test.py::test_broken_projections_in_backups_1 | https://s3.amazonaws.com/clickhouse-test-reports/66344/1c92e093319d474dba22cd36eaf2284a12d1fb64/integration_tests__aarch64__[1_6].html |
| 0 | 60c0957ab | 2.47 | 2024-07-11 18:38:49 | 10 days and 15 hours | Integration tests (aarch64) [1/6] | test_broken_projections/test.py::test_broken_projections_in_backups_1 | https://s3.amazonaws.com/clickhouse-test-reports/0/60c0957abd24c5038647b4e42b2da82481136102/integration_tests__aarch64__[1_6].html |
| 64889 | 4195ce218 | 2.5 | 2024-07-05 23:02:32 | 16 days and 11 hours | Integration tests (aarch64) [1/6] | test_broken_projections/test.py::test_broken_projections_in_backups_1 | https://s3.amazonaws.com/clickhouse-test-reports/64889/4195ce21836306d15a45dba85e8b2cdf0958aadf/integration_tests__aarch64__[1_6].html |
| 66012 | 073471530 | 2.72 | 2024-07-02 23:26:15 | 19 days and 10 hours | Integration tests (aarch64) [1/6] | test_broken_projections/test.py::test_broken_projections_in_backups_1 | https://s3.amazonaws.com/clickhouse-test-reports/66012/073471530b1e6bc8f08b959b4071cd0a376f24e1/integration_tests__aarch64__[1_6].html |
| 0 | 2b686fa8a | 2.25 | 2024-07-02 13:15:58 | 19 days and 21 hours | Integration tests (aarch64) [1/6] | test_broken_projections/test.py::test_broken_projections_in_backups_1 | https://s3.amazonaws.com/clickhouse-test-reports/0/2b686fa8ade5c5360b05c58c3416e6dfb17a27e5/integration_tests__aarch64__[1_6].html |
| 0 | a13a5efb7 | 2.24 | 2024-06-27 10:30:54 | 24 days and 23 hours | Integration tests (aarch64) [1/6] | test_broken_projections/test.py::test_broken_projections_in_backups_1 | https://s3.amazonaws.com/clickhouse-test-reports/0/a13a5efb7180f1f61f658f27e0e70907503dcede/integration_tests__aarch64__[1_6].html |



</details>


Fails with error:

```
E           helpers.client.QueryRuntimeException: Client failed! Return code: 233, stderr: Received exception from server (version 24.7.1):
E           Code: 1001. DB::Exception: Received from 172.16.3.5:9000. DB::Exception: std::__1::__fs::filesystem::filesystem_error: filesystem error: in file_size: No such file or directory ["/var/lib/clickhouse/store/d6d/d6d04e68-fae3-4ff5-95a9-053b86ff1461/all_2_2_0/proj1.proj/c.bin"]. Stack trace:
E           
E           0. std::system_error::system_error(std::error_code, String const&) @ 0x0000000015962000
E           1. std::__fs::filesystem::filesystem_error::filesystem_error[abi:v15007](String const&, std::__fs::filesystem::path const&, std::error_code) @ 0x000000000c910bd0
E           2. void std::__fs::filesystem::__throw_filesystem_error[abi:v15007]<String&, std::__fs::filesystem::path const&, std::error_code const&>(String&, std::__fs::filesystem::path const&, std::error_code const&) @ 0x000000001591bd2c
E           3. std::__fs::filesystem::detail::(anonymous namespace)::ErrorHandler<unsigned long>::report(std::error_code const&) const (.llvm.12967353655292048837) @ 0x0000000015920ba0
E           4. std::__fs::filesystem::__file_size(std::__fs::filesystem::path const&, std::error_code*) @ 0x0000000015920a00
E           5. DB::DiskLocal::getFileSize(String const&) const @ 0x000000000fa38a4c
E           6. DB::BackupWriterDisk::copyFileFromDisk(String const&, std::shared_ptr<DB::IDisk>, String const&, bool, unsigned long, unsigned long) @ 0x000000000f4f2548
E           7. DB::BackupImpl::writeFile(DB::BackupFileInfo const&, std::shared_ptr<DB::IBackupEntry const>) @ 0x000000000f4d7374
E           8. DB::BackupsWorker::writeBackupEntries(std::shared_ptr<DB::IBackup>, std::vector<std::pair<String, std::shared_ptr<DB::IBackupEntry const>>, std::allocator<std::pair<String, std::shared_ptr<DB::IBackupEntry const>>>>&&, String const&, std::shared_ptr<DB::IBackupCoordination>, bool, std::shared_ptr<DB::QueryStatus>)::$_0::operator()() const @ 0x000000000f458a78
E           9. DB::ThreadPoolCallbackRunnerLocal<void, ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>, std::function<void ()>>::operator()(std::function<void ()>&&, Priority)::'lambda'()::operator()() @ 0x000000000ef96be0
E           10. std::packaged_task<void ()>::operator()() @ 0x000000000cd71198
E           11. ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::worker(std::__list_iterator<ThreadFromGlobalPoolImpl<false, true>, void*>) @ 0x000000000c96ab64
E           12. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false, true>::ThreadFromGlobalPoolImpl<void ThreadPoolImpl<ThreadFromGlobalPoolImpl<false, true>>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>(void&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000c96e730
E           13. void* std::__thread_proxy[abi:v15007]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000c96d3e0
E           14. ? @ 0x000000000007d5c8
E           15. ? @ 0x00000000000e5edc
E           . (STD_EXCEPTION)
E           (query: backup table test4 to Disk('backups', 'b1-hRkNky9U') settings check_projection_parts=false;)

```

--- a/src/Backups/BackupsWorker.cpp
+++ b/src/Backups/BackupsWorker.cpp
@@ -384,6 +384,7 @@ BackupsWorker::BackupsWorker(ContextMutablePtr global_context, size_t num_backup
     , allow_concurrent_backups(global_context->getConfigRef().getBool("backups.allow_concurrent_backups", true))
     , allow_concurrent_restores(global_context->getConfigRef().getBool("backups.allow_concurrent_restores", true))
     , remove_backup_files_after_failure(global_context->getConfigRef().getBool("backups.remove_backup_files_after_failure", true))
+    , test_randomize_order(global_context->getConfigRef().getBool("backups.test_randomize_order", false))
     , test_inject_sleep(global_context->getConfigRef().getBool("backups.test_inject_sleep", false))
     , log(getLogger("BackupsWorker"))
     , backup_log(global_context->getBackupLog())
@@ -713,14 +714,25 @@ void BackupsWorker::writeBackupEntries(
     bool always_single_threaded = !backup->supportsWritingInMultipleThreads();
     auto & thread_pool = getThreadPool(ThreadPoolId::BACKUP_COPY_FILES);
 
+    std::vector<size_t> writing_order;
+    if (test_randomize_order)
+    {
+        /// Randomize the order in which we write backup entries to the backup.
+        writing_order.resize(backup_entries.size());
+        std::iota(writing_order.begin(), writing_order.end(), 0);
+        std::shuffle(writing_order.begin(), writing_order.end(), thread_local_rng);
+    }
+
     ThreadPoolCallbackRunnerLocal<void> runner(thread_pool, "BackupWorker");
     for (size_t i = 0; i != backup_entries.size(); ++i)
     {
         if (failed)
             break;
 
-        auto & entry = backup_entries[i].second;
-        const auto & file_info = file_infos[i];
+        size_t index = !writing_order.empty() ? writing_order[i] : i;
+
+        auto & entry = backup_entries[index].second;
+        const auto & file_info = file_infos[index];
 
         auto job = [&]()
         {

--- a/src/Backups/BackupsWorker.h
+++ b/src/Backups/BackupsWorker.h
@@ -119,6 +119,7 @@ private:
     const bool allow_concurrent_backups;
     const bool allow_concurrent_restores;
     const bool remove_backup_files_after_failure;
+    const bool test_randomize_order;
     const bool test_inject_sleep;
 
     LoggerPtr log;

--- a/tests/integration/test_broken_projections/config.d/backups.xml
+++ b/tests/integration/test_broken_projections/config.d/backups.xml
@@ -9,5 +9,6 @@
     </storage_configuration>
     <backups>
         <allowed_disk>backups</allowed_disk>
+        <test_randomize_order>true</test_randomize_order>
     </backups>
 </clickhouse>

--- a/tests/integration/test_broken_projections/test.py
+++ b/tests/integration/test_broken_projections/test.py
@@ -148,23 +148,22 @@ def break_part(node, table, part):
     bash(node, f"rm '{part_path}/columns.txt'")
 
 
-def get_broken_projections_info(node, table, active=True):
+def get_broken_projections_info(node, table, part=None, projection=None, active=True):
+    parent_name_filter = f" AND parent_name = '{part}'" if part else ""
+    name_filter = f" AND name = '{projection}'" if projection else ""
     return node.query(
         f"""
-    SELECT parent_name, name, errors.name FROM
-    (
-        SELECT parent_name, name, exception_code
+    SELECT parent_name, name, exception
         FROM system.projection_parts
         WHERE table='{table}'
         AND database=currentDatabase()
         AND is_broken = 1
         AND active = {active}
-    ) AS parts_info
-    INNER JOIN system.errors AS errors
-    ON parts_info.exception_code = errors.code
+        {parent_name_filter}
+        {name_filter}
     ORDER BY parent_name, name
     """
-    ).strip()
+    )
 
 
 def get_projections_info(node, table):
@@ -312,8 +311,8 @@ def test_broken_ignored(cluster):
 
     # Projection 'proj1' from part all_2_2_0 will now appear in broken parts info
     # because it was marked broken during "check table" query.
-    assert "all_2_2_0\tproj1\tFILE_DOESNT_EXIST" in get_broken_projections_info(
-        node, table_name
+    assert "FILE_DOESNT_EXIST" in get_broken_projections_info(
+        node, table_name, part="all_2_2_0", projection="proj1"
     )
 
     # Check table query will also show a list of parts which have broken projections.
@@ -323,14 +322,14 @@ def test_broken_ignored(cluster):
     break_projection(node, table_name, "proj2", "all_2_2_0", "data")
 
     # It will not yet appear in broken projections info.
-    assert "proj2" not in get_broken_projections_info(node, table_name)
+    assert not get_broken_projections_info(node, table_name, projection="proj2")
 
     # Select now fails with error "File doesn't exist"
     check(node, table_name, 0, "proj2", "FILE_DOESNT_EXIST")
 
     # Projection 'proj2' from part all_2_2_0 will now appear in broken parts info.
-    assert "all_2_2_0\tproj2\tNO_FILE_IN_DATA_PART" in get_broken_projections_info(
-        node, table_name
+    assert "NO_FILE_IN_DATA_PART" in get_broken_projections_info(
+        node, table_name, part="all_2_2_0", projection="proj2"
     )
 
     # Second select works, because projection is now marked as broken.
@@ -340,7 +339,7 @@ def test_broken_ignored(cluster):
     break_projection(node, table_name, "proj2", "all_3_3_0", "data")
 
     # It will not yet appear in broken projections info.
-    assert "all_3_3_0" not in get_broken_projections_info(node, table_name)
+    assert not get_broken_projections_info(node, table_name, part="all_3_3_0")
 
     insert(node, table_name, 20, 5)
     insert(node, table_name, 25, 5)
@@ -371,8 +370,8 @@ def test_broken_ignored(cluster):
         node, table_name
     )
 
-    assert "all_3_3_0" in get_broken_projections_info(node, table_name, active=False)
-    assert "all_2_2_0" in get_broken_projections_info(node, table_name, active=True)
+    assert get_broken_projections_info(node, table_name, part="all_3_3_0", active=False)
+    assert get_broken_projections_info(node, table_name, part="all_2_2_0", active=True)
 
     # 0 because of all_2_2_0
     check(node, table_name, 0)
@@ -396,8 +395,8 @@ def test_materialize_broken_projection(cluster):
     break_projection(node, table_name, "proj1", "all_1_1_0", "metadata")
     reattach(node, table_name)
 
-    assert "all_1_1_0\tproj1\tNO_FILE_IN_DATA_PART" in get_broken_projections_info(
-        node, table_name
+    assert "NO_FILE_IN_DATA_PART" in get_broken_projections_info(
+        node, table_name, part="all_1_1_0", projection="proj1"
     )
     assert "Part all_1_1_0 has a broken projection proj1" in check_table_full(
         node, table_name
@@ -406,8 +405,8 @@ def test_materialize_broken_projection(cluster):
     break_projection(node, table_name, "proj2", "all_1_1_0", "data")
     reattach(node, table_name)
 
-    assert "all_1_1_0\tproj2\tFILE_DOESNT_EXIST" in get_broken_projections_info(
-        node, table_name
+    assert "FILE_DOESNT_EXIST" in get_broken_projections_info(
+        node, table_name, part="all_1_1_0", projection="proj2"
     )
     assert "Part all_1_1_0 has a broken projection proj2" in check_table_full(
         node, table_name
@@ -469,8 +468,8 @@ def test_broken_projections_in_backups_2(cluster):
     break_projection(node, table_name, "proj2", "all_2_2_0", "part")
     check(node, table_name, 0, "proj2", "ErrnoException")
 
-    assert "all_2_2_0\tproj2\tFILE_DOESNT_EXIST" == get_broken_projections_info(
-        node, table_name
+    assert "FILE_DOESNT_EXIST" in get_broken_projections_info(
+        node, table_name, part="all_2_2_0", projection="proj2"
     )
 
     assert "FILE_DOESNT_EXIST" in node.query_and_get_error(
@@ -524,8 +523,8 @@ def test_broken_projections_in_backups_3(cluster):
     assert "Part all_1_1_0 has a broken projection proj1" in check_table_full(
         node, table_name
     )
-    assert "all_1_1_0\tproj1\tFILE_DOESNT_EXIST" == get_broken_projections_info(
-        node, table_name
+    assert "FILE_DOESNT_EXIST" in get_broken_projections_info(
+        node, table_name, part="all_1_1_0", projection="proj1"
     )
 
     backup_name = f"b4-{get_random_string()}"
@@ -545,8 +544,11 @@ def test_broken_projections_in_backups_3(cluster):
     )
 
     check(node, table_name, 0)
-    assert "all_1_1_0\tproj1\tNO_FILE_IN_DATA_PART" == get_broken_projections_info(
-        node, table_name
+    assert (
+        "Projection directory proj1.proj does not exist while loading projections"
+        in get_broken_projections_info(
+            node, table_name, part="all_1_1_0", projection="proj1"
+        )
     )
 
 
@@ -569,7 +571,7 @@ def test_check_part_thread(cluster):
     break_projection(node, table_name, "proj2", "all_2_2_0", "data")
 
     # It will not yet appear in broken projections info.
-    assert "proj2" not in get_broken_projections_info(node, table_name)
+    assert not get_broken_projections_info(node, table_name, projection="proj2")
 
     # Select now fails with error "File doesn't exist"
     check(node, table_name, 0, "proj2", "FILE_DOESNT_EXIST", do_check_command=False)
@@ -606,15 +608,15 @@ def test_broken_on_start(cluster):
     break_projection(node, table_name, "proj2", "all_2_2_0", "data")
 
     # It will not yet appear in broken projections info.
-    assert "proj2" not in get_broken_projections_info(node, table_name)
+    assert not get_broken_projections_info(node, table_name, projection="proj2")
 
     # Select now fails with error "File doesn't exist"
     # We will mark projection as broken.
     check(node, table_name, 0, "proj2", "FILE_DOESNT_EXIST")
 
     # Projection 'proj2' from part all_2_2_0 will now appear in broken parts info.
-    assert "all_2_2_0\tproj2\tNO_FILE_IN_DATA_PART" in get_broken_projections_info(
-        node, table_name
+    assert "NO_FILE_IN_DATA_PART" in get_broken_projections_info(
+        node, table_name, part="all_2_2_0", projection="proj2"
     )
 
     # Second select works, because projection is now marked as broken.
@@ -623,7 +625,7 @@ def test_broken_on_start(cluster):
     node.restart_clickhouse()
 
     # It will not yet appear in broken projections info.
-    assert "proj2" in get_broken_projections_info(node, table_name)
+    assert get_broken_projections_info(node, table_name, projection="proj2")
 
     # Select works
     check(node, table_name, 0)
@@ -654,7 +656,7 @@ def test_mutation_with_broken_projection(cluster):
         node, table_name
     )
 
-    assert "" == get_broken_projections_info(node, table_name)
+    assert not get_broken_projections_info(node, table_name)
 
     check(node, table_name, 1)
 
@@ -662,21 +664,21 @@ def test_mutation_with_broken_projection(cluster):
     break_projection(node, table_name, "proj2", "all_2_2_0_4", "data")
 
     # It will not yet appear in broken projections info.
-    assert "proj2" not in get_broken_projections_info(node, table_name)
+    assert not get_broken_projections_info(node, table_name, projection="proj2")
 
     # Select now fails with error "File doesn't exist"
     # We will mark projection as broken.
     check(node, table_name, 0, "proj2", "FILE_DOESNT_EXIST")
 
     # Projection 'proj2' from part all_2_2_0_4 will now appear in broken parts info.
-    assert "all_2_2_0_4\tproj2\tNO_FILE_IN_DATA_PART" in get_broken_projections_info(
-        node, table_name
+    assert "NO_FILE_IN_DATA_PART" in get_broken_projections_info(
+        node, table_name, part="all_2_2_0_4", projection="proj2"
     )
 
     # Second select works, because projection is now marked as broken.
     check(node, table_name, 0)
 
-    assert "all_2_2_0_4" in get_broken_projections_info(node, table_name)
+    assert get_broken_projections_info(node, table_name, part="all_2_2_0_4")
 
     node.query(
         f"ALTER TABLE {table_name} DELETE WHERE _part == 'all_0_0_0_4' SETTINGS mutations_sync = 1"
@@ -690,14 +692,10 @@ def test_mutation_with_broken_projection(cluster):
 
     # Still broken because it was hardlinked.
     broken = get_broken_projections_info(node, table_name)
-    assert (
-        "all_2_2_0_5" in broken or "" == broken
-    )  # second could be because of a merge.
+    if broken:  # can be not broken because of a merge.
+        assert get_broken_projections_info(node, table_name, part="all_2_2_0_5")
 
-    if "" == broken:
-        check(node, table_name, 1)
-    else:
-        check(node, table_name, 0)
+    check(node, table_name, not broken)
 
     node.query(
         f"ALTER TABLE {table_name} DELETE WHERE c == 13 SETTINGS mutations_sync = 1"
@@ -710,6 +708,6 @@ def test_mutation_with_broken_projection(cluster):
     )
 
     # Not broken anymore.
-    assert "" == get_broken_projections_info(node, table_name)
+    assert not get_broken_projections_info(node, table_name)
 
     check(node, table_name, 1)

--- a/tests/integration/test_broken_projections/test.py
+++ b/tests/integration/test_broken_projections/test.py
@@ -450,54 +450,6 @@ def get_random_string(string_length=8):
     return "".join((random.choice(alphabet) for _ in range(string_length)))
 
 
-def test_broken_projections_in_backups_1(cluster):
-    node = cluster.instances["node"]
-
-    table_name = "test4"
-    create_table(node, table_name, 1, aggressive_merge=False, data_prefix=table_name)
-
-    node.query("SYSTEM STOP MERGES")
-
-    insert(node, table_name, 0, 5)
-    insert(node, table_name, 5, 5)
-    insert(node, table_name, 10, 5)
-    insert(node, table_name, 15, 5)
-
-    assert ["all_0_0_0", "all_1_1_0", "all_2_2_0", "all_3_3_0"] == get_parts(
-        node, table_name
-    )
-
-    check(node, table_name, 1)
-
-    break_projection(node, table_name, "proj1", "all_2_2_0", "data")
-    check(node, table_name, 0, "proj1", "FILE_DOESNT_EXIST")
-
-    assert "all_2_2_0\tproj1\tNO_FILE_IN_DATA_PART" in get_broken_projections_info(
-        node, table_name
-    )
-
-    backup_name = f"b1-{get_random_string()}"
-    assert "BACKUP_CREATED" in node.query(
-        f"""
-    set backup_restore_keeper_fault_injection_probability=0.0;
-    backup table {table_name} to Disk('backups', '{backup_name}') settings check_projection_parts=false;
-    """
-    )
-
-    assert "RESTORED" in node.query(
-        f"""
-    drop table {table_name} sync;
-    set backup_restore_keeper_fault_injection_probability=0.0;
-    restore table {table_name} from Disk('backups', '{backup_name}');
-    """
-    )
-
-    node.query("SYSTEM STOP MERGES")
-
-    check(node, table_name, 1)
-    assert "" == get_broken_projections_info(node, table_name)
-
-
 def test_broken_projections_in_backups_2(cluster):
     node = cluster.instances["node"]
 

